x/posts/types/models.go

```diff
@@ -0,0 +1,669 @@
+package types
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/gogo/protobuf/proto"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// ParsePostID parses the given value as a post id, returning an error if it's invalid
+func ParsePostID(value string) (uint64, error) {
+	if value == "" {
+		return 0, nil
+	}
+
+	postID, err := strconv.ParseUint(value, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("invalid post id: %s", err)
+	}
+	return postID, nil
+}
+
+// NewPost allows to build a new Post instance
+func NewPost(
+	subspaceID uint64,
+	id uint64,
+	externalID string,
+	text string,
+	author string,
+	conversationID uint64,
+	entities *Entities,
+	referencedPosts []PostReference,
+	replySetting ReplySetting,
+	creationDate time.Time,
+	lastEditedDate *time.Time,
+) Post {
+	return Post{
+		SubspaceID:      subspaceID,
+		ID:              id,
+		ExternalID:      externalID,
+		Text:            text,
+		Entities:        entities,
+		Author:          author,
+		ConversationID:  conversationID,
+		ReferencedPosts: referencedPosts,
+		ReplySettings:   replySetting,
+		CreationDate:    creationDate,
+		LastEditedDate:  lastEditedDate,
+	}
+}
+
+// Validate implements fmt.Validator
+func (p Post) Validate() error {
+	if p.SubspaceID == 0 {
+		return fmt.Errorf("invalid subspace id: %d", p.SubspaceID)
+	}
+
+	if p.ID == 0 {
+		return fmt.Errorf("invalid post id: %d", p.ID)
+	}
+
+	if p.Entities != nil {
+		err := p.Entities.Validate()
+		if err != nil {
+			return fmt.Errorf("invalid entities: %s", err)
+		}
+
+		// Make sure that no entity has a start or end index that is greater to the text length
+		maxIndexAllowed := uint64(0)
+		if len(strings.TrimSpace(p.Text)) > 0 {
+			maxIndexAllowed = uint64(len(strings.TrimSpace(p.Text)) - 1)
+		}
+
```

Comment:
It is missing to validate the content of entities. Currently the following post can pass the validation:
```json
{
    ...
    "text": "Hello #Desmos",
    "entities": [{
        "hash_tags": [
            {
                "start": 0,
                "end": 3,
                "tag": "desmos"
            }
        ]
    }]
}
```

Reply:
This is wanted. I believe tags should not have any validation because they might be used for some particular cases. 

As an example, let's suppose I want to tag you into a post. I would then write: 
```
Hi @dadamu! Welcome!
```

At this post, the tag would be something like
```
{
  "start": 3,
  "end": 9,
  "tag": "desmos1....."
}
```

As you can see, there is no way we can validate this properly. I prefer to leave it to the applications to handle these cases freely without any constrictions